### PR TITLE
Handle Compilation Errors & Runtime mappings for empty regexp

### LINKED ISSUE #39794: [Task]: Provide a compile-time error for empty regexp
### Description

Provide a compile-time error for ```re `` ```.

### Describe your task(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

#39281, #39444

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #39795: [Task]: Handle regex patterns initialised with empty strings
### Description

Consider following examples

```
public function main() {
    string a = "";
    string:RegExp b = re `${""}`;
    string:RegExp c = re `${a}`;
}
```

Regular expressions that initialised with empty strings like in above example should be convert into non capturing groups in Ballerina runtime.

ex:-
The above two regex patterns will convert into `?:` in Ballerina runtime.

The current behaviour is , regular expressions that initialised with empty strings like in above example are treated as empty strings in Ballerina runtime.

### Describe your task(s)

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by sanjana**
Fixes #39794
Fixes #39795

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java [hunk 1]
```diff
@@ -38,6 +39,10 @@
 public class Find {
 
     public static BArray find(BRegexpValue regExp, BString str, long startIndex) {
+        if (isEmptyRegexp(regExp)) {
```
Comment (SasinduDilshara) [inline langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:42]: We can remove this logic in this PR

Comment (sanjana) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:42]: Will be removing this with #39795 once the Runtime case is handled as well.

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 2]
```diff
@@ -1958,3 +1958,6 @@ error.worker.send.action.not.allowed.in.lock.statement=\
 
 error.worker.receive.action.not.allowed.in.lock.statement=\
   using receive action within the lock statement is not allowed to prevent possible deadlocks
+
+error.empty.regexp.string.disallowed=\
+  regular expression with empty string is not allowed
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:1963]:
```suggestion
  regular expression is not allowed: empty RegExp
```

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java [hunk 3]
```diff
@@ -61,11 +61,21 @@ static Matcher getMatcher(BRegexpValue regexpVal, BString inputStr) {
     static Matcher getMatcher(BRegexpValue regexpVal, String inputStr) {
         // Map the required ballerina regexp constructs to java.
         RegExpValue translatedRegExpVal = RegExpFactory.translateRegExpConstructs((RegExpValue) regexpVal);
-        String patternStr = StringUtils.getStringValue(translatedRegExpVal, null);
+        String patternStr = getRegexpStrValue(translatedRegExpVal);
         Pattern pattern = Pattern.compile(patternStr);
         return pattern.matcher(inputStr);
     }
 
+    private static String getRegexpStrValue(RegExpValue translatedRegExpVal) {
```
Comment (pcnfernando) [inline langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java:69]: Don't do it here. Do it at `translateRegExpConstructs()`

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 2 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39848/commits/927c1a6378676f68f309748de84c9bb66f76f2f9.

File: langlib/langlib-test/src/test/resources/test-src/regexp_empty_test_negative.bal [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    _ = regexp:findAll(re ``, "There once was a king who liked to sing");
+    _ = regexp:findAll(re ``, "");
+    _ = regexp:findGroups(re ``, "Butter was bought by Betty but the butter was bitter");
+    _ = regexp:findGroups(re ``, "");
+    _ = regexp:findAllGroups(re ``, "rubble, trouble, bubble, hubble");
+    _ = regexp:findAllGroups(re ``, "");
+    _ = regexp:fullMatchGroups(re = re ``, str = "HelloWorld");
+    _ = regexp:fullMatchGroups(re = re ``, str = "");
+    _ = regexp:matchAt(re = re ``, str = "HelloWorld");
+    _ = regexp:matchAt(re = re ``, str = "HelloWorld", startIndex = 4);
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/regexp_empty_test_negative.bal:32]: Shall we add a one test like

```
var a = re ``.find("a", 1);
```

**Review by SasinduDilshara — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 5]
```diff
@@ -1958,3 +1958,6 @@ error.worker.send.action.not.allowed.in.lock.statement=\
 
 error.worker.receive.action.not.allowed.in.lock.statement=\
   using receive action within the lock statement is not allowed to prevent possible deadlocks
+
+error.empty.regexp.string.disallowed=\
+  regular expression is not allowed: empty RegExp
```
Comment (SasinduDilshara) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:1963]: Shall we update the error message to something like `Empty regular expressions are not allowed`.
wdyt?

**Review by SasinduDilshara — CHANGES_REQUESTED** [review]

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java (hunk 3 above)
Comment (sanjana) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java:69]: @pcnfernando do we need to have the `.toString()` and `.toBalString()` methods return the non-capturing group as well?

```
string:RegExp r = re `${""}`;
r.toBalString(); // output: re `(?:)`
```

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 5 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:1963]: @pcnfernando compared to https://github.com/ballerina-platform/ballerina-lang/pull/39848#discussion_r1134970386, WDYT?

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java (hunk 3 above)
Comment (sanjana) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39848/commits/26145dd1c98d0a23840715c7b3dd693760c2f5bc.

File: langlib/langlib-test/src/test/resources/test-src/regexp_empty_test_negative.bal (hunk 4 above)
Comment (sanjana) [reply langlib/langlib-test/src/test/resources/test-src/regexp_empty_test_negative.bal:32]: Added in https://github.com/ballerina-platform/ballerina-lang/pull/39848/commits/26145dd1c98d0a23840715c7b3dd693760c2f5bc.

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 5 above)
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:1963]: ok

**Review by SasinduDilshara — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 6]
```diff
@@ -1256,6 +1258,18 @@ private void checkSupportedConfigType(BVarSymbol varSymbol, Location location, S
         }
     }
 
+    @Override
+    public void visit(BLangRegExpTemplateLiteral node, AnalyzerData data) {
+        analyzeNode(node.reDisjunction, data);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1263]: Do the empty regexp validation here

**Pull request merged**